[amazonechocontrol] Websocket error after out of memory OH2 crash

Get websocket error of amazon binding and after a while out of memory.

https://community.openhab.org/t/web-socket-error/100553

+1 Just for completeness:

Normally I get the following exception:
```
2020-06-13 22:53:22.202 [INFO ] [control.internal.WebSocketConnection] - Web Socket error
java.nio.channels.AsynchronousCloseException: null
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.close(HttpConnectionOverHTTP.java:181) [bundleFile:9.4.20.v20190813]
	at java.util.ArrayList.forEach(ArrayList.java:1540) [?:?]
	at org.eclipse.jetty.client.AbstractConnectionPool.close(AbstractConnectionPool.java:208) [bundleFile:9.4.20.v20190813]
	at org.eclipse.jetty.client.DuplexConnectionPool.close(DuplexConnectionPool.java:237) [bundleFile:9.4.20.v20190813]
	at org.eclipse.jetty.client.HttpDestination.close(HttpDestination.java:385) [bundleFile:9.4.20.v20190813]
	at org.eclipse.jetty.client.HttpClient.doStop(HttpClient.java:260) [bundleFile:9.4.20.v20190813]
	at org.eclipse.jetty.util.component.AbstractLifeCycle.stop(AbstractLifeCycle.java:93) [bundleFile:9.4.20.v20190813]
	at org.eclipse.jetty.util.component.ContainerLifeCycle.stop(ContainerLifeCycle.java:180) [bundleFile:9.4.20.v20190813]
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStop(ContainerLifeCycle.java:201) [bundleFile:9.4.20.v20190813]
	at org.eclipse.jetty.websocket.client.WebSocketClient.doStop(WebSocketClient.java:371) [bundleFile:9.4.20.v20190813]
	at org.eclipse.jetty.util.component.AbstractLifeCycle.stop(AbstractLifeCycle.java:93) [bundleFile:9.4.20.v20190813]
	at org.openhab.binding.amazonechocontrol.internal.WebSocketConnection.close(WebSocketConnection.java:159) [bundleFile:?]
	at org.openhab.binding.amazonechocontrol.internal.WebSocketConnection$2.run(WebSocketConnection.java:184) [bundleFile:?]
	at java.util.TimerThread.mainLoop(Timer.java:556) [?:?]
	at java.util.TimerThread.run(Timer.java:506) [?:?]
```

But as others also report OH crashes as well sometimes. In that case, i see the following exception.
```
2020-06-13 22:55:03.188 [INFO ] [control.internal.WebSocketConnection] - Web Socket error
org.eclipse.jetty.io.EofException: null
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:283) ~[?:?]
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.flush(SslConnection.java:910) ~[?:?]
	at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:422) ~[?:?]
	at org.eclipse.jetty.io.WriteFlusher.completeWrite(WriteFlusher.java:378) ~[?:?]
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.lambda$fill$1(SslConnection.java:669) ~[?:?]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:782) [bundleFile:9.4.20.v20190813]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:918) [bundleFile:9.4.20.v20190813]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.io.IOException: Broken pipe
	at sun.nio.ch.FileDispatcherImpl.write0(Native Method) ~[?:?]
	at sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:47) ~[?:?]
	at sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:113) ~[?:?]
	at sun.nio.ch.IOUtil.write(IOUtil.java:58) ~[?:?]
	at sun.nio.ch.IOUtil.write(IOUtil.java:50) ~[?:?]
	at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:466) ~[?:?]
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:261) ~[?:?]
	
```

--- a/bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java
+++ b/bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java
@@ -33,8 +33,6 @@ import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.eclipse.jdt.annotation.Nullable;
 import org.eclipse.jetty.util.ssl.SslContextFactory;
 import org.eclipse.jetty.websocket.api.Session;
-import org.eclipse.jetty.websocket.api.UpgradeRequest;
-import org.eclipse.jetty.websocket.api.UpgradeResponse;
 import org.eclipse.jetty.websocket.api.annotations.OnWebSocketClose;
 import org.eclipse.jetty.websocket.api.annotations.OnWebSocketConnect;
 import org.eclipse.jetty.websocket.api.annotations.OnWebSocketError;
@@ -42,7 +40,6 @@ import org.eclipse.jetty.websocket.api.annotations.OnWebSocketMessage;
 import org.eclipse.jetty.websocket.api.annotations.WebSocket;
 import org.eclipse.jetty.websocket.client.ClientUpgradeRequest;
 import org.eclipse.jetty.websocket.client.WebSocketClient;
-import org.eclipse.jetty.websocket.client.io.UpgradeListener;
 import org.openhab.binding.amazonechocontrol.internal.jsons.JsonPushCommand;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -119,8 +116,7 @@ public class WebSocketConnection {
 
             initPongTimeoutTimer();
 
-            sessionFuture = webSocketClient.connect(amazonEchoControlWebSocket, uri, request,
-                    amazonEchoControlWebSocket);
+            sessionFuture = webSocketClient.connect(amazonEchoControlWebSocket, uri, request);
         } catch (URISyntaxException e) {
             logger.debug("Initialize web socket failed", e);
         }

--- a/bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java
+++ b/bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java
@@ -35,7 +35,11 @@ import org.eclipse.jetty.util.ssl.SslContextFactory;
 import org.eclipse.jetty.websocket.api.Session;
 import org.eclipse.jetty.websocket.api.UpgradeRequest;
 import org.eclipse.jetty.websocket.api.UpgradeResponse;
-import org.eclipse.jetty.websocket.api.WebSocketListener;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketClose;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketConnect;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketError;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketMessage;
+import org.eclipse.jetty.websocket.api.annotations.WebSocket;
 import org.eclipse.jetty.websocket.client.ClientUpgradeRequest;
 import org.eclipse.jetty.websocket.client.WebSocketClient;
 import org.eclipse.jetty.websocket.client.io.UpgradeListener;
@@ -60,7 +64,7 @@ public class WebSocketConnection {
     private final Gson gson = new Gson();
     private final WebSocketClient webSocketClient;
     private final IWebSocketCommandHandler webSocketCommandHandler;
-    private final Listener listener;
+    private final AmazonEchoControlWebSocket amazonEchoControlWebSocket;
 
     private @Nullable Session session;
     private @Nullable Timer pingTimer;
@@ -72,11 +76,10 @@ public class WebSocketConnection {
     public WebSocketConnection(String amazonSite, List<HttpCookie> sessionCookies,
             IWebSocketCommandHandler webSocketCommandHandler) throws IOException {
         this.webSocketCommandHandler = webSocketCommandHandler;
-        listener = new Listener();
+        amazonEchoControlWebSocket = new AmazonEchoControlWebSocket();
 
         SslContextFactory sslContextFactory = new SslContextFactory();
         webSocketClient = new WebSocketClient(sslContextFactory);
-
         try {
             String host;
             if (StringUtils.equalsIgnoreCase(amazonSite, "amazon.com")) {
@@ -110,16 +113,14 @@ public class WebSocketConnection {
             }
 
             ClientUpgradeRequest request = new ClientUpgradeRequest();
-            request.setHeader("host", host);
-            request.setHeader("Cache-Control", "no-cache");
-            request.setHeader("Pragma", "no-cache");
+            request.setHeader("Host", host);
             request.setHeader("Origin", "alexa." + amazonSite);
-
             request.setCookies(cookiesForWs);
 
             initPongTimeoutTimer();
-            sessionFuture = webSocketClient.connect(listener, uri, request);
 
+            sessionFuture = webSocketClient.connect(amazonEchoControlWebSocket, uri, request,
+                    amazonEchoControlWebSocket);
         } catch (URISyntaxException e) {
             logger.debug("Initialize web socket failed", e);
         }
@@ -134,7 +135,7 @@ public class WebSocketConnection {
 
             @Override
             public void run() {
-                listener.sendPing();
+                amazonEchoControlWebSocket.sendPing();
             }
         }, 180000, 180000);
     }
@@ -156,7 +157,7 @@ public class WebSocketConnection {
             try {
                 session.close();
             } catch (Exception e) {
-                logger.debug("Closing sessing failed", e);
+                logger.debug("Closing session failed", e);
             }
         }
         logger.trace("Connect future = {}", sessionFuture);
@@ -200,13 +201,13 @@ public class WebSocketConnection {
         }, 60000);
     }
 
-    class Listener implements WebSocketListener, UpgradeListener {
-        private final Logger logger = LoggerFactory.getLogger(Listener.class);
+    @WebSocket(maxTextMessageSize = 64 * 1024, maxBinaryMessageSize = 64 * 1024)
+    @SuppressWarnings("unused")
+    public class AmazonEchoControlWebSocket {
         int msgCounter = -1;
         int messageId;
 
-        Listener() {
-            logger.trace("Initialized listener");
+        AmazonEchoControlWebSocket() {
             this.messageId = ThreadLocalRandom.current().nextInt(0, Short.MAX_VALUE);
         }
 
@@ -377,7 +378,7 @@ public class WebSocketConnection {
             return message;
         }
 
-        @Override
+        @OnWebSocketConnect
         public void onWebSocketConnect(@Nullable Session session) {
             if (session != null) {
                 this.msgCounter = -1;
@@ -388,7 +389,7 @@ public class WebSocketConnection {
             }
         }
 
-        @Override
+        @OnWebSocketMessage
         public void onWebSocketBinary(byte @Nullable [] data, int offset, int len) {
             if (data == null) {
                 return;
@@ -427,35 +428,23 @@ public class WebSocketConnection {
             }
         }
 
-        @Override
+        @OnWebSocketMessage
         public void onWebSocketText(@Nullable String message) {
+            logger.trace("Received text message: '{}'", message);
         }
 
-        @Override
+        @OnWebSocketClose
         public void onWebSocketClose(int code, @Nullable String reason) {
             logger.info("Web Socket close {}. Reason: {}", code, reason);
             WebSocketConnection.this.close();
         }
 
-        @Override
+        @OnWebSocketError
         public void onWebSocketError(@Nullable Throwable error) {
             logger.info("Web Socket error", error);
             if (!closed) {
                 WebSocketConnection.this.close();
             }
-        }
-
-        @Override
-        public void onHandshakeResponse(@Nullable UpgradeResponse response) {
-            if (response == null) {
-                logger.trace("Received null response in onHandshakeResponse");
-                return;
-            }
-            logger.trace("UpgradeResponse: {}", response);
-        }
-
-        @Override
-        public void onHandshakeRequest(@Nullable UpgradeRequest request) {
         }
 
         public void sendPing() {

--- a/bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java
+++ b/bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java
@@ -25,6 +25,7 @@ import java.util.List;
 import java.util.Timer;
 import java.util.TimerTask;
 import java.util.UUID;
+import java.util.concurrent.Future;
 import java.util.concurrent.ThreadLocalRandom;
 
 import org.apache.commons.lang.StringUtils;
@@ -32,9 +33,12 @@ import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.eclipse.jdt.annotation.Nullable;
 import org.eclipse.jetty.util.ssl.SslContextFactory;
 import org.eclipse.jetty.websocket.api.Session;
+import org.eclipse.jetty.websocket.api.UpgradeRequest;
+import org.eclipse.jetty.websocket.api.UpgradeResponse;
 import org.eclipse.jetty.websocket.api.WebSocketListener;
 import org.eclipse.jetty.websocket.client.ClientUpgradeRequest;
 import org.eclipse.jetty.websocket.client.WebSocketClient;
+import org.eclipse.jetty.websocket.client.io.UpgradeListener;
 import org.openhab.binding.amazonechocontrol.internal.jsons.JsonPushCommand;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -64,6 +68,9 @@ public class WebSocketConnection {
     Listener listener;
     boolean closed;
     IWebSocketCommandHandler webSocketCommandHandler;
+    private long messageCounter = 0;
+    private long messageReceived = 0;
+    private @Nullable Future<Session> future;
 
     public WebSocketConnection(String amazonSite, List<HttpCookie> sessionCookies,
             IWebSocketCommandHandler webSocketCommandHandler) throws IOException {
@@ -114,7 +121,7 @@ public class WebSocketConnection {
             request.setCookies(cookiesForWs);
 
             initPongTimeoutTimer();
-            webSocketClient.connect(listener, uri, request);
+            future = webSocketClient.connect(listener, uri, request);
 
         } catch (URISyntaxException e) {
             logger.debug("Initialize web socket failed", e);
@@ -155,8 +162,14 @@ public class WebSocketConnection {
                 logger.debug("Closing sessing failed", e);
             }
         }
+        logger.trace("Connect future = {}", future);
+        if (!future.isDone()) {
+            future.cancel(true);
+        }
         try {
-            webSocketClient.stop();
+            if (webSocketClient.isStarted()) {
+                webSocketClient.stop();
+            }
         } catch (InterruptedException e) {
             // Just ignore
         } catch (Exception e) {
@@ -169,6 +182,7 @@ public class WebSocketConnection {
         Timer pongTimeoutTimer = this.pongTimeoutTimer;
         this.pongTimeoutTimer = null;
         if (pongTimeoutTimer != null) {
+            logger.trace("Cancelling pong timeout");
             pongTimeoutTimer.cancel();
         }
     }
@@ -177,20 +191,24 @@ public class WebSocketConnection {
         clearPongTimeoutTimer();
         Timer pongTimeoutTimer = new Timer();
         this.pongTimeoutTimer = pongTimeoutTimer;
+        logger.trace("Scheduling pong timeout");
         pongTimeoutTimer.schedule(new TimerTask() {
 
             @Override
             public void run() {
+                logger.trace("Pong timeout reached. Closing connection.");
                 close();
             }
         }, 60000);
     }
 
-    class Listener implements WebSocketListener {
+    class Listener implements WebSocketListener, UpgradeListener {
+        private final Logger logger = LoggerFactory.getLogger(Listener.class);
         int msgCounter = -1;
         int messageId;
 
         Listener() {
+            logger.trace("Initialized listener");
             this.messageId = ThreadLocalRandom.current().nextInt(0, Short.MAX_VALUE);
         }
 
@@ -203,6 +221,7 @@ public class WebSocketConnection {
         }
 
         void sendMessage(byte[] buffer) {
+            WebSocketConnection.this.messageCounter++;
             try {
                 logger.debug("Send message with length {}", buffer.length);
                 Session session = WebSocketConnection.this.session;
@@ -374,6 +393,7 @@ public class WebSocketConnection {
 
         @Override
         public void onWebSocketBinary(byte @Nullable [] data, int offset, int len) {
+            WebSocketConnection.this.messageReceived++;
             if (data == null) {
                 return;
             }
@@ -424,7 +444,22 @@ public class WebSocketConnection {
         @Override
         public void onWebSocketError(@Nullable Throwable error) {
             logger.info("Web Socket error", error);
-            WebSocketConnection.this.close();
+            if (!closed) {
+                WebSocketConnection.this.close();
+            }
+        }
+
+        @Override
+        public void onHandshakeResponse(@Nullable UpgradeResponse response) {
+            if (response == null) {
+                logger.trace("Received null response in onHandshakeResponse");
+                return;
+            }
+            logger.trace("UpgradeResponse: {}", response);
+        }
+
+        @Override
+        public void onHandshakeRequest(@Nullable UpgradeRequest request) {
         }
 
         public void sendPing() {

--- a/bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java
+++ b/bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java
@@ -58,19 +58,16 @@ import com.google.gson.JsonSyntaxException;
 public class WebSocketConnection {
     private final Logger logger = LoggerFactory.getLogger(WebSocketConnection.class);
     private final Gson gson = new Gson();
-    WebSocketClient webSocketClient;
-    @Nullable
-    Session session;
-    @Nullable
-    Timer pingTimer;
-    @Nullable
-    Timer pongTimeoutTimer;
-    Listener listener;
-    boolean closed;
-    IWebSocketCommandHandler webSocketCommandHandler;
-    private long messageCounter = 0;
-    private long messageReceived = 0;
-    private @Nullable Future<Session> future;
+    private final WebSocketClient webSocketClient;
+    private final IWebSocketCommandHandler webSocketCommandHandler;
+    private final Listener listener;
+
+    private @Nullable Session session;
+    private @Nullable Timer pingTimer;
+    private @Nullable Timer pongTimeoutTimer;
+    private @Nullable Future<?> sessionFuture;
+
+    private boolean closed;
 
     public WebSocketConnection(String amazonSite, List<HttpCookie> sessionCookies,
             IWebSocketCommandHandler webSocketCommandHandler) throws IOException {
@@ -121,7 +118,7 @@ public class WebSocketConnection {
             request.setCookies(cookiesForWs);
 
             initPongTimeoutTimer();
-            future = webSocketClient.connect(listener, uri, request);
+            sessionFuture = webSocketClient.connect(listener, uri, request);
 
         } catch (URISyntaxException e) {
             logger.debug("Initialize web socket failed", e);
@@ -162,9 +159,10 @@ public class WebSocketConnection {
                 logger.debug("Closing sessing failed", e);
             }
         }
-        logger.trace("Connect future = {}", future);
-        if (!future.isDone()) {
-            future.cancel(true);
+        logger.trace("Connect future = {}", sessionFuture);
+        final Future<?> sessionFuture = this.sessionFuture;
+        if (!sessionFuture.isDone()) {
+            sessionFuture.cancel(true);
         }
         try {
             if (webSocketClient.isStarted()) {
@@ -221,7 +219,6 @@ public class WebSocketConnection {
         }
 
         void sendMessage(byte[] buffer) {
-            WebSocketConnection.this.messageCounter++;
             try {
                 logger.debug("Send message with length {}", buffer.length);
                 Session session = WebSocketConnection.this.session;
@@ -393,7 +390,6 @@ public class WebSocketConnection {
 
         @Override
         public void onWebSocketBinary(byte @Nullable [] data, int offset, int len) {
-            WebSocketConnection.this.messageReceived++;
             if (data == null) {
                 return;
             }

